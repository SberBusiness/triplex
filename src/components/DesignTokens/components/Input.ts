import {TDesignTokenValue, TDesignTokenValues} from '@sberbusiness/triplex/components/DesignTokens/types/DesignTokenTypes';

// Название токенов компонента Input.
export const designTokensComponentsInputKeys = [
    'Background_Default',
    'Background_Disabled',
    'BorderColor_Default',
    'BorderColor_Disabled',
    'BorderColor_Error',
    'BorderColor_Focus',
    'Caret_Color',
    'Color_Default',
    'Color_Disabled',
    'Placeholder_Color',
] as const;
// Тип, содержащий названия токенов компонента Input.
export type TDesignTokensComponentsInputKeys = (typeof designTokensComponentsInputKeys)[number];
// Тип, содержащий названия токенов компонента Input и их значения.
export type TDesignTokensComponentsInputValue = Record<TDesignTokensComponentsInputKeys, TDesignTokenValue>;
// Тип, содержащий названия токенов компонента Input и их значения в светлой и темной теме.
export type TDesignTokensComponentsInputValues = Record<TDesignTokensComponentsInputKeys, TDesignTokenValues>;
// Тип локальных токенов компонента Input.
export type TDesignTokensComponentsInput = {Input: TDesignTokensComponentsInputValue};

// Токены компонента Input в светлой и темной темах.
export const Input_Tokens: TDesignTokensComponentsInputValues = {
    Background_Default: [{ref: 'Basic.100'}, {ref: 'Basic.900'}], // var(--triplex-Input-Background_Default)
    Background_Disabled: [{ref: 'Neutral.100'}, {ref: 'Basic.700'}], // var(--triplex-Input-Background_Disabled)
    BorderColor_Default: [{ref: 'Neutral.500'}, {ref: 'Neutral.900'}], // var(--triplex-Input-BorderColor_Default)
    BorderColor_Disabled: [{ref: 'Neutral.500'}, {ref: 'Basic.300'}], // var(--triplex-Input-BorderColor_Disabled)
    BorderColor_Error: [{ref: 'Error.700'}, {ref: 'Error.500'}], // var(--triplex-Input-BorderColor_Error)
    BorderColor_Focus: [{ref: 'Basic.700'}, {ref: 'Neutral.700'}], // var(--triplex-Input-BorderColor_Focus)
    Caret_Color: [{ref: 'Basic.700'}, {ref: 'Neutral.100'}], // var(--triplex-Input-Caret_Color)
    Color_Default: [{ref: 'Basic.700'}, {ref: 'Neutral.100'}], // var(--triplex-Input-Color_Default)
    Color_Disabled: [{ref: 'Neutral.900'}, {ref: 'Basic.300'}], // var(--triplex-Input-Color_Disabled)
    Placeholder_Color: [{ref: 'Neutral.900'}, {ref: 'Neutral.900'}], // var(--triplex-Input-Placeholder_Color)
};
