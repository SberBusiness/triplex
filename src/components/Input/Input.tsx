import React from 'react';
import {EInputGroupPosition} from '@sberbusiness/triplex/components/InputGroup/InputGroup';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';

/** Свойства компонента "Поле для ввода информации". */
export interface IInputProps extends React.InputHTMLAttributes<HTMLInputElement> {
    /** Состояние ошибки. */
    error?: boolean;
    /** Позиция внутри компонента InputGroup. */
    groupPosition?: EInputGroupPosition;
}

const mapInputGroupPositionToCSSClass = {
    [EInputGroupPosition.LEFT]: 'cssClass[left]',
    [EInputGroupPosition.INTERMEDIATE]: 'cssClass[intermediate]',
    [EInputGroupPosition.RIGHT]: 'cssClass[right]',
};

/** Компонент "Поле для ввода информации". */
export const Input = React.forwardRef<HTMLInputElement, IInputProps>(({className, groupPosition, error, ...rest}, ref) => {
    const classNames = classnames(
        'cssClass[input]',
        {'cssClass[grouped]': !!groupPosition},
        groupPosition && mapInputGroupPositionToCSSClass[groupPosition],
        {'cssClass[error]': !!error},
        className
    );

    return <input className={classNames} {...rest} ref={ref} data-tx={process.env.npm_package_version} />;
});

Input.displayName = 'Input';
