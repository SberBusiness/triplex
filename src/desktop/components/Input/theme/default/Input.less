@import '../../../../common/theme/default';

.input {
    .input-base();

    width: 100%;

    &::-ms-clear {
        display: none;
    }
}

.grouped {
    flex: 1;

    &:not(:first-child) {
        margin-left: -1px;
    }

    &.left {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
    }

    &.intermediate {
        border-radius: 0;
    }

    &.right {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
    }

    &.error {
        z-index: 1;
    }

    &:focus {
        z-index: 2;
    }
}
