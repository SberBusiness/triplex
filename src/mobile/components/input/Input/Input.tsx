import {Ref} from '@sberbusiness/triplex/common/types/Ref';
import {classnames} from '@sberbusiness/triplex/common/utils/classnames/classnames';
import * as React from 'react';

/**
 * Свойства компонента.
 */
export interface IInputProps extends React.InputHTMLAttributes<HTMLInputElement> {
    getRef?: Ref<HTMLInputElement>;
    error?: boolean;
}

/**
 * Компонент Input.
 */
export const Input: React.FC<IInputProps> = ({className, error, getRef, ...HTMLInputAttributes}) => (
    <input
        {...HTMLInputAttributes}
        type="text"
        className={classnames(className, 'cssClass[mobileInput]', {'cssClass[error]': Boolean(error)})}
        ref={getRef}
    />
);

Input.displayName = 'Input';
