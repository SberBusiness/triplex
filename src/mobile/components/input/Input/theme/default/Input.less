@import '../../../../../common/theme/default';

.mobileInput {
    .input-basic();

    width: 100%;
}
