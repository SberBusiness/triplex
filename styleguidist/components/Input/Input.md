```jsx
<Input placeholder="Введите значение" />
```

### Disabled state

```jsx
<Input placeholder="Введите значение" disabled />
```

### Error state

```jsx
<Input placeholder="Введите значение" error />
```
