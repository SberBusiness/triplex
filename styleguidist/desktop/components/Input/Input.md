```jsx
import {ExampleControlPanel} from '../common/ExampleControlPanel/ExampleControlPanel';

const [value, setValue] = React.useState('');
const [disabled, setDisabled] = React.useState(false);
const [error, setError] = React.useState(false);

const renderControlPanel = () => (
    <ExampleControlPanel>
        <input
            type="checkbox"
            checked={disabled}
            onChange={(event) => setDisabled(event.target.checked)}
            data-label="Disabled"
        />
        <input
            type="checkbox"
            checked={error}
            onChange={(event) => setError(event.target.checked)}
            data-label="Error"
        />
    </ExampleControlPanel>
);

<>
    {renderControlPanel()}  
    <Input
        value={value}
        onChange={(event) => setValue(event.target.value)}
        placeholder="Введите значение"
        disabled={disabled}
        error={error}
    />
</>
```