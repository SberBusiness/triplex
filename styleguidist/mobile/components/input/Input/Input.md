```jsx
import React, { useState } from 'react';
import {CheckboxYGroup} from '@sberbusiness/triplex/desktop/components/Checkbox/CheckboxYGroup';
import {Checkbox} from '@sberbusiness/triplex/desktop/components/Checkbox/Checkbox';
import {MobileOptionsWrapper} from '../../common/MobileOptionsWrapper/MobileOptionsWrapper';

const [state, setState] = useState({
    disabled: false,
    error: false,
    value: ''
});

const options = [
    {
        label: 'Disabled',
        checked: state.disabled,
        onChange: (e) => setState({...state, disabled: e.target.checked})
    }, {
        label: 'Error',
        checked: state.error,
        onChange: (e) => setState({...state, error: e.target.checked})
    }
];

<>
    <MobileOptionsWrapper>
        <CheckboxYGroup>
            {options.map((item, index) => (
                <Checkbox key={index} onChange={item.onChange}>{item.label}</Checkbox>
            ))}
        </CheckboxYGroup>
    </MobileOptionsWrapper>

    <Input 
        disabled={state.disabled} 
        error={state.error} 
        value={state.value} 
        onChange={(e) => setState({...state, value: e.target.value})} 
        placeholder={'Введите значение'}
    />
</>
```