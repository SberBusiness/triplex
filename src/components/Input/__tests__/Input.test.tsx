import React from 'react';
import {Input} from '../Input';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';

describe('Input', () => {
    beforeEach(() => {
        allure.feature('Input');
    });

    it('renders correctly', () => {
        const onChangeMock = jest.fn();
        const tree = renderer.create(<Input onChange={onChangeMock} value={'default-value'} />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should call onChange prop', () => {
        const handleChange = jest.fn();
        const event = {
            currentTarget: {value: 'some-value'},
        };
        const component = shallow(<Input onChange={handleChange} value={'default-value'} />);
        component.simulate('change', event);

        expect(handleChange).toHaveBeenCalledWith({currentTarget: {value: 'some-value'}});
    });
});
