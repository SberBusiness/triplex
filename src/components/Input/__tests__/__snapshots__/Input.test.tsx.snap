// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
<input
  className="cssClass[input]"
  data-tx="10.0.0"
  onChange={[MockFunction]}
  value="default-value"
/>
`;
